dotnet/src/Agents/OpenAI/Internal/AssistantThreadActions.cs

```diff
@@ -301,18 +304,52 @@ async Task PollRunStatusAsync()
 
             do
             {
-                // Reduce polling frequency after a couple attempts
-                await Task.Delay(agent.PollingOptions.GetPollingInterval(count), cancellationToken).ConfigureAwait(false);
-                ++count;
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (count > 0)
+                {
+                    // Reduce polling frequency after a couple attempts
+                    await Task.Delay(agent.PollingOptions.GetPollingInterval(count), cancellationToken).ConfigureAwait(false);
+                    ++count;
+                }
 
 #pragma warning disable CA1031 // Do not catch general exception types
                 try
                 {
                     run = await client.GetRunAsync(threadId, run.Id, cancellationToken).ConfigureAwait(false);
                 }
+                catch (ClientResultException clientException) when (clientException.Status > 0)
+                {
+                    // The presence of a `Status` code means the server responded with error
+                    throw;
+                }
+                catch (AggregateException aggregateException) when (aggregateException.InnerException is ClientResultException innerClientException)
+                {
+                    // The presence of a `Status` code means the server responded with error
+                    if (innerClientException.Status > 0)
+                    {
+                        throw;
+                    }
+
+                    // Check maximum retry count
+                    if (count >= agent.PollingOptions.MaximumRetryCount)
+                    {
+                        throw;
+                    }
+
+                    // Retry for potential transient failure
+                    continue;
+                }
                 catch
```

Comment:
Do we want to retry on any exception here or should this be scoped to ClientResultException where status > 0

Reply:
Updated based on offline discussion